themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

```diff
@@ -0,0 +1,205 @@
+`Register Resource` request to the Pulumi engine.
+
+No matter what the actual state of your infrastructure is, running your program will always send to intended
+resource model to the engine.
+
+Our previous example was shown in Python, but recently we also delivered [support for YAML](https://www.pulumi.com/docs/intro/languages/yaml/).
+The creation of our S3 bucket could be converted to this snippet:
+
+```yaml
+resources:
+  bucket:
+    type: aws:s3:Bucket
+```
+
+The separation of language support from the engine allows us to offer both imperative & declarative solutions.
+But remember that none of the supported languages do any of the provisioning directly in the language host.
+
+If the resource provisioning is not taking place in the language host, where is the magic happening then?
+
+### CLI and Engine
+
+In the previous step, you found out that the language host sends requests to the engine to fullfil your to be
+infrastructure.
+
+It is now that our Pulumi engine gets to work. The engine combines the intended model of the infrastructure
+received from the language host, the current state recorded in the state backend and the actual resource state
+to compute which actions need to be executed to bring the actual state in line with the intended model.
+
+Our little example contains dependencies:
+every `s3.BucketObject` uses the `bucket.id` as way to define in which bucket these objects should be stored.
+The property `id` from the `s3.Bucket` is an `Output`. Outputs are Pulumi's way of tracking which property of one resource
+is required by another, hereby creating a dependency between the resources. Our engine uses all these outputs
+passed from one resource to another as the vertices in a directed acyclic graph (DAG). The engine determines
+the order of actions based on this graph.
+
+On a first run of `pulumi up` of our example program, our engine will first create the bucket, wait for
+the provisioning to be complete, after which the actual bucket id is passed to the creation of
+all the bucket objects. Since none of the bucket objects depend on other resources, these bucket
+objects can all be provisioned concurrently. The engine will
+
+On a second run, assuming no modifications to our example program, the Pulumi engine will compare the to be model
+with the actual state and conclude that nothing needs to be done.
+
+Let's crank up the nr of bucket objects to 11.
+
+```python
+from pulumi_aws import s3
+
+bucket = s3.Bucket('bucket')
+
+for i in range(11): # <- nr of objects increased by 1
+    s3.BucketObject(
+        f'object-{i}',
+        s3.BucketObjectArgs(
+            bucket=bucket.id,
+            key=str(i),
+        )
+    )
+```
+
+The third Pulumi run will find the bucket and the first 10 objects in the state. The only action taken
+now is the creation of the 11th bucket object. This brings the actual state back in sync with your
```

Comment:
Would it be a good idea to capture the snippet from `pulumi preview` that shows that only the 11th object will be created?

Reply:
Done.